### PR TITLE
Update e2e testing nodePort service listening on same port but different protocols

### LINKED ISSUE #81418: Update "service using same port but different protocol" e2e to test service reachability
**What would you like to be added**:
Following are two e2e testing service functioning over same port but different protocol, but do not test service reachability.
1. should be able to update NodePorts with two same port numbers but different protocols
2. should use same NodePort with same port but different protocols
**Note**: _Both the e2e are verifying the same behavior of "service using same port but different protocols"_ differs in `service update` and `service create only`.

**_Primary goal:_** Test service reachability for these E2Es.
**_Secondary goal:_** As both the above e2e are verifying same E2E behavior, it's good to have single e2e in spite of 2 differing in "updating" and "create only" nodePort service.
For "create only", suite has another specific e2e `should be able to create a functioning NodePort service`.

**Why is this needed**:
E2E Improvement and make e2e eligible for Conformance.
Service related e2e are getting promoted in #77865

/area conformance
/sig testing
@kubernetes/cncf-conformance-wg
/assign @mgdevstack

### PR DESCRIPTION
**Pull request by mgdevstack**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test

/kind feature
> /kind flake

**What this PR does / why we need it**:
1. Updates `should be able to update NodePorts with two same port numbers but different protocols` to test service reachability and improved e2e Name and behavior.
2. Removes `should use same NodePort with same port but different protocols` as this behaviour would be tested by above updated e2e.

**Which issue(s) this PR fixes**:

Fixes #81418

**Special notes for your reviewer**:
Changes are made to make e2e eligible for Conformance.
Source of Issue : https://github.com/kubernetes/kubernetes/pull/77865#issuecomment-492849235

**Does this PR introduce a user-facing change?**:

```release-note
None
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```
/area conformance
/sig testing
@kubernetes/cncf-conformance-wg
@kubernetes/sig-node-pr-reviews
@kubernetes/sig-architecture-pr-reviews

### CODE REVIEW

**Comment by mgdevstack** [comment]
Once merged, we can promote this e2e to Conformance in this release cycle v1.16 which is very near.
/priority important-soon

**Review by mattjmcnaughton — APPROVED** [review]
/lgtm

Probably a good idea to have someone more involved with the conformance work review this as well, but it looks good to me :)

File: test/e2e/network/service.go
Comment (oomichi) [inline test/e2e/network/service.go]: nit: Here can be `framework.ExpectEqual(len(nodePortService.Spec.Ports), 2)`

File: test/e2e/framework/service/jig.go
Comment (oomichi) [inline test/e2e/framework/service/jig.go]: nit: I am wondering we can merge this new function into the above existing function because the difference is just this command here. But at this time, it is fine and we can take care later.

**Review by oomichi — COMMENTED** [review]
/lgtm

File: test/e2e/framework/service/jig.go
Comment (mgdevstack) [reply test/e2e/framework/service/jig.go]: Thanks for the suggestion, I was in the same mindset to have single function and leverage flags as args but to keep PR simple, I have not introduced here. As you mentioned, I can create another PR to refactor it.

Following is raw code.
```
var ServeHostNameImageDefaultFlags = map[string]string{
	HTTP: "--http",
	TCP:  "--tcp",
	UDP:  "--udp",
}


func enableServeHostNameImageCmdFlags(enableFlags ...string) (enabledFlags []string){
	gomega.Expect(len(enableFlags)).NotTo(gomega.BeNumerically(">", 2), "HTTP only or TCP/UDP requests can be served but not both simultaneously")
	var flags = make(map[string]string,len(ServeHostNameImageDefaultFlags))
	for _,f := range enableFlags{
		switch f{
		case TCP, UDP:
			framework.ExpectNotEqual(flags[f],ServeHostNameImageDefaultFlags[HTTP], "HTTP flag can not be enabled along with TCP/UDP")
			flags[f] = "--http=false"
			flags[f] = ServeHostNameImageDefaultFlags[f]
		case HTTP:
			if flags[TCP] != "" || flags[UDP] != ""{
				framework.Failf("TCP/UDP flag is enabled so HTTP flag can not be enabled")
			}
			flags[f] = ServeHostNameImageDefaultFlags[f]
		default:
			framework.Failf("Unexpected protocol flags are passed. Valid protocol flags are HTTP, TCP and UDP")
		}
	}
	for _, f := range flags{
                enabledFlags = append(enableFlags, f)
	}
    return
}

flags := enableServeHostNameImageCmdFlags(TCP, UDP)
```

File: test/e2e/network/service.go
Comment (mgdevstack) [inline test/e2e/network/service.go]: nit: ```framework.ExpectNotEqual(port.NodePort, 0, errorMessage...)```

Comment (mgdevstack) [reply test/e2e/network/service.go]: Done.

File: test/e2e/network/service.go
Comment (mgdevstack) [reply test/e2e/network/service.go]: Done.

**Comment by oomichi** [comment]
Thanks for updating

/lgtm

**Comment by mgdevstack** [comment]
Failing e2e in `pull-kubernetes-conformance-kind-ipv6 ` are not related to this PR.

**Comment by oomichi** [comment]
Thanks for doing this.
The following `/hold` is to free merge pool bandwidth for 1.16-targeted PRs as https://github.com/kubernetes/kubernetes/pull/82108#issuecomment-526207592
We need to wait for making these high priority PRs merged first.
(still 3 such PRs are on the queue)

/approve
/hold

**Comment by oomichi** [comment]
as https://kubernetes.slack.com/archives/C2C40FMNF/p1567190475094100

/hold cancel

**Comment by oomichi** [comment]
That seems unrelated test failure.

/test pull-kubernetes-e2e-aks-engine-azure-windows

**Comment by adelina-t** [comment]
> That seems unrelated test failure.
> 
> /test pull-kubernetes-e2e-aks-engine-azure-windows

Yes, it is unrelated. Because of an image change for redis, that is unsupported by windows, the guestbook application cannot run, thus the test will fail. It was supposed to be removed from windows e2e presubmit, but seems it was an oversight that I shall fix ASAP.

I ran the aks-engine-azure-windows job on this PR because I wanted to see if the changes don't impact Windows.

**Comment by mgdevstack** [comment]
E2E non-related to this PR are failing on Azure job which requires fix mentioned in https://github.com/kubernetes/kubernetes/pull/81419#issuecomment-528242299 and milestone is not added for this release i.e. 1.16 so holding this PR to stop repeated retest.
/hold

**Comment by johnbelamaric** [comment]
@mgdevstack please rebase and resolve conflicts

File: test/e2e/framework/service/jig.go
Comment (alejandrox1) [reply test/e2e/framework/service/jig.go]: @mgdevstack could you please create a follow up issue for this, so we dont forget?
once we get that we can cancel the hold on this.

**Review by alejandrox1 — COMMENTED** [review]
/lgtm
/milestone v1.17

**Comment by mgdevstack** [comment]
/hold cancel
As per https://github.com/kubernetes/kubernetes/pull/81419#discussion_r329366521 tracking issue under #83347

**Pull request merged**
